api/odilo.py

```diff
@@ -0,0 +1,495 @@
+            if exception_on_401:
+                # This is our second try. Give up.
+                raise Exception("Something's wrong with the patron OAuth Bearer Token!")
+            else:
+                # Refresh the token and try again.
+                self.check_creds(True)
+                return self.patron_request(patron, pin, url, extra_headers, data, True)
+        else:
+            return response
+
+    def get_patron_credential(self, patron, pin):
+        """Create an OAuth token for the given patron."""
+
+        def refresh(credential):
+            return self.get_patron_access_token(credential, patron, pin)
+
+        return Credential.lookup(self._db, DataSource.ODILO, "OAuth Token", patron, refresh)
+
+    def get_patron_access_token(self, credential, patron, pin):
+        """Request an OAuth bearer token that allows us to act on
+        behalf of a specific patron.
+        """
+
+        self.client_key = patron
+        self.client_secret = pin
+        message = self.refresh_creds(credential)
+
+        if 'OK' == message:
+            return credential
+        else:
+            raise PatronAuthorizationFailedException(message)
+
+    def checkout(self, patron, pin, licensepool, internal_format):
+        """Check out a book on behalf of a patron.
+
+        :param patron: a Patron object for the patron who wants
+        to check out the book.
+
+        :param pin: The patron's alleged password.
+
+        :param licensepool: Identifier of the book to be checked out is 
+        attached to this licensepool.
+
+        :param internal_format: Represents the patron's desired book format.
+
+        :return: a LoanInfo object.
+        """
+        record_id = licensepool.identifier.identifier
+
+        # We have 2 formats, 'EBOOK_STREAMING' could be in our API ('PDF', 'EPUB'):
+        ebook_streaming = False
+        if internal_format == 'EBOOK_STREAMING':
+            ebook_streaming = True
+            internal_format = 'EPUB'  # Try first ebook_streaming as EPUB
+
+        # Data just as 'x-www-form-urlencoded', no JSON
+        payload = dict(patronId=patron, format=internal_format)
+
+        response = self.patron_request(
+            patron, pin, self.CHECKOUT_ENDPOINT.replace("{recordId}", record_id),
+            extra_headers={'Content-Type': 'application/x-www-form-urlencoded'},
+            data=payload)
+
+        if response.content:
+            response_json = response.json()
+            if response.status_code == 404:
+                error_id, error_description = self.raise_exception_on_error(response_json)
+                raise CannotLoan('Error id: %s, description: %s' % (error_id, error_description))
+            else:
+                return self.loan_info_from_odilo_checkout(licensepool.collection, response_json)
+
+        elif response.status_code == 400 and ebook_streaming:
+            internal_format = 'PDF'  # Retry now with ebook_streaming as PDF
```

Comment:
If you decide to keep going down this path, I think you should check the error message to make sure switching to PDF is really the solution. But maybe this is covered in your TODO on line 161.